### PR TITLE
Support dynamic shaped arrays

### PR DESCRIPTION
**Pull request by mavenlin**
## Description

Make `Spec<dtype>` to support `Container<dtype> = std::unique_ptr<TArray<dtype>>`.
The corresponding state/action is converted to a `np.array` of `np.array` on python side.

## Motivation and Context

In certain cases, the state/action of the environment is a dynamic shaped array each time, e.g. sequence of natural language. Currently, only the `num_players` can be configured to be dynamic. With this PR, we can now support dynamic shape of state/action for each env in the batch.

- [ ] I have raised an issue to propose this change ([required](https://envpool.readthedocs.io/en/latest/pages/contributing.html) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Put an `x` in all the boxes that apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] New environment (non-breaking change which adds 3rd-party environment)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of example)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://envpool.readthedocs.io/en/latest/pages/contributing.html) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [x] I have updated the documentation accordingly.
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the code using `make lint` (**required**)
- [x] I have ensured `make bazel-test` pass. (**required**)

### CODE REVIEW

File: envpool/core/array.h [hunk 1]
```diff
@@ -198,4 +220,10 @@ class Array {
   [[nodiscard]] std::shared_ptr<char> SharedPtr() const { return ptr_; }
 };
 
+template <typename Dtype>
+class TArray : public Array {
```
Comment (Trinkle23897) [inline envpool/core/array.h:224]: Do we really need TArray at this point?

**Comment by Trinkle23897** [comment]
Another thought: we only enable using dynamic shape on Recv function (c++->python), is it possible to do it on Send function (python->c++)?

File: envpool/core/array.h (hunk 1 above)
Comment (mavenlin) [reply envpool/core/array.h:224]: You're right, it can still work without TArray.

**Comment by mavenlin** [comment]
It is possible. It will take some specialization of the NumpyToArray template to convert from np array of array to our `Container` type.

Only problem is the destruction. As our array is not type aware, it will not perform destruction of its elements when gets freed.

> Another thought: we only enable using dynamic shape on Recv function (c++->python), is it possible to do it on Send function (python->c++)?

**Comment by Trinkle23897** [comment]
Should we temporarily ban it and document the reason or ...?

**Review by Trinkle23897 — APPROVED** [review]

**Pull request merged**